cupyx/scipy/spatial/_kdtree_utils.py

```diff
@@ -0,0 +1,918 @@
+        }
+
+        if(p == CUDART_INF) {
+            dist = max(dist, diff);
+        } else {
+            dist = min(dist, diff);
+        }
+    }
+    return dist;
+}
+
+template<typename T>
+__device__ double compute_distance(
+        const T* __restrict__ point1, const T* __restrict__ point2,
+        const double* __restrict__ box_bounds,
+        const int n_dims, const double p, const int stride,
+        const bool take_root) {
+
+    if(abs(p) == CUDART_INF) {
+        return compute_distance_inf<T>(
+            point1, point2, box_bounds, n_dims, p, stride);
+    }
+
+    double dist = 0.0;
+    for(int i = 0; i < n_dims; i++) {
+        double diff = abs(point1[i] - point2[i * stride]);
+        double dim_bound = box_bounds[i];
+        if(diff > dim_bound - diff) {
+            diff = dim_bound - diff;
+        }
+        dist += pow(diff, p);
+    }
+
+    if(take_root) {
+        dist = pow(dist, 1.0 / p);
+    }
+    return dist;
+}
+
+template<typename T>
+__device__ T insort(
+        const long long curr, const T dist, const int k, const int n,
+        T* distances, long long* nodes, bool check) {
+
+    if(check && dist > distances[k - 1]) {
+        return distances[k - 1];
+    }
+
+    long long left = 0;
+    long long right = k - 1;
+
+    while(left != right) {
+        long long pos = (left + right) / 2;
+        if(distances[pos] < dist) {
+            left = pos + 1;
+        } else {
+            right = pos;
+        }
+    }
+
+    long long node_to_insert = curr;
+    T dist_to_insert = dist;
+    T dist_to_return = dist;
+
+    for(long long i = left; i < k; i++) {
+        long long node_tmp = nodes[i];
+        T dist_tmp = distances[i];
+
+        nodes[i] = node_to_insert;
+        distances[i] = dist_to_insert;
+
+        dist_to_return = max(dist_to_return, distances[i]);
+        node_to_insert = node_tmp;
+        dist_to_insert = dist_tmp;
+
+    }
+
+    return dist_to_return;
+}
+
+template<typename T>
+__device__ double min_bound_dist(
+        const T* __restrict__ point_bounds, const T point_dim,
+        const double dim_bound, const int dim) {
+    const T min_bound = point_bounds[0];
+    const T max_bound = point_bounds[1];
+
+    double min_dist = abs(min_bound - point_dim);
+    min_dist = min(min_dist, dim_bound - min_dist);
+
+    double max_dist = abs(max_bound - point_dim);
+    max_dist = min(max_dist, dim_bound - max_dist);
+    return min(min_dist, max_dist);
+}
+
+template<typename T>
+__device__ void compute_knn(
+        const int k, const int n, const int n_dims, const double eps,
+        const double p, const double dist_bound, const bool periodic,
+        const T* __restrict__ point, const T* __restrict__ tree,
+        const long long* __restrict__ index,
+        const double* __restrict__ box_bounds,
+        const T* __restrict__ tree_bounds,
+        double* distances, long long* nodes) {
+
+    volatile long long prev = -1;
+    volatile long long curr = 0;
```

Comment:
Does `volatile` keyword yields random results in Windows?

Reply:
Let me try to remove these volatile keywords.